### PR TITLE
Enhance InjectAgent to Handle Only HTTPGet, TCP Readiness Probes

### LINKED ISSUE #3909: Kserve agent exec probe failed
I attempted to add a readiness probe to the InferenceService YAML configuration. Specifically, I used the following configuration:

```
readinessProbe: 
  exec:
    command:
      - /bin/sh
      - '-c'
      - >-
        curl -X GET
        0.0.0.0:8080/v2/models/model/versions/1/ready
```

After applying the InferenceService YAML, KServe created a pod named predictor. This pod contains two containers: kserve-container and agent. The readiness probe was successfully applied to the kserve-container, but the agent container raised an error: "exec probe not supported".

I would like to understand how to ensure that the readiness probe is applied only to the kserve-container and not to the agent container. Is there a way to achieve this configuration, or is there an alternative approach for implementing readiness probes in KServe?

KServe Version: 13.0.1

### PR DESCRIPTION
**Pull request by LOADBC**
### What this PR does / why we need it

This PR enhances the `InjectAgent` function to handle readiness probes more effectively:
- It  injects only `HTTPGet` and `TCPSocket` readiness probes into the agent container.
- It skips `Exec` readiness probes, as they are incompatible with the agent container's readiness checks.
- The changes prevent potential errors and ensure compatibility with Kubernetes best practices for readiness probes.

### Which issue(s) this PR fixes

This PR addresses issues where non-`HTTPGet` and `Exec` readiness probes cause the agent container to fail.

Fixes: #3909
Feature: #3989

### Type of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

### Feature/Issue validation/testing

**Test A:** Verified that `HTTPGet` readiness probes are correctly inherited and serialized.
**Test B:** Verified that `Exec` readiness probes are skipped, with appropriate logging.
**Test C:** Verified that `TCPSocket` readiness probes are inherited and serialized correctly.

#### Logs

`Running tool: /usr/local/go/bin/go test -timeout 30s -coverprofile=/tmp/vscode-go0yOou5/go-code-cover `github.com/kserve/kserve/pkg/webhook/admission/pod`

`ok  	github.com/kserve/kserve/pkg/webhook/admission/pod	6.261s	coverage: 87.8% of statements`

### Special notes for your reviewer

- Please confirm that if this PR changes any image versions, it only changes those.

### Checklist

- [ ] Have you added unit/e2e tests that prove your fix is effective or that this feature works?
- [x] Has code been commented, particularly in hard-to-understand areas?
- [ ] Have you made corresponding changes to the documentation?

### Release note

- Add any necessary release notes here.

### Re-running failed tests

- `/rerun-all` - rerun all failed workflows.
- `/rerun-workflow <workflow name>` - rerun a specific failed workflow.

### CODE REVIEW

**Comment by LOADBC** [comment]
Hi @spolti,

In my previous PR, I encountered some issues with my forked repository, which required me to close it. I've now resolved those issues and created a fresh PR with the same changes. Please take a look when you have a moment.

File: pkg/webhook/admission/pod/agent_injector.go [hunk 1]
```diff
@@ -238,12 +238,35 @@ func (ag *AgentInjector) InjectAgent(pod *v1.Pod) error {
...
 	if !queueProxyAvailable {
-		readinessProbeJson, err := json.Marshal(pod.Spec.Containers[0].ReadinessProbe)
-		if err != nil {
-			return err
+		readinessProbe := pod.Spec.Containers[0].ReadinessProbe
+	
+		// Check if the readiness probe exists
+		if readinessProbe != nil {
+			if readinessProbe.HTTPGet != nil {
+				// Marshal the HTTPGet readiness probe into JSON format
```
Comment (spolti) [inline pkg/webhook/admission/pod/agent_injector.go]: maybe this is cleaner:

```go
	if readinessProbe.HTTPGet != nil || readinessProbe.TCPSocket != nil {
		// Marshal the HTTPGet readiness probe into JSON format
		readinessProbeJson, err := json.Marshal(readinessProbe)
		if err != nil {
			return fmt.Errorf("failed to marshal readiness probe: %w", err)
		}

		// Append the marshaled readiness probe as an environment variable for the agent container
		agentEnvs = append(agentEnvs, v1.EnvVar{Name: "SERVING_READINESS_PROBE", Value: string(readinessProbeJson)})
	} else if readinessProbe.Exec != nil {
		// Skip ExecAction probes; do not inherit them
		fmt.Printf("INFO: Exec readiness probe skipped for pod %s/%s\n", pod.Namespace, pod.Name)
	}
```

Comment (LOADBC) [reply pkg/webhook/admission/pod/agent_injector.go]: Thanks, @spolti  I’ve incorporated your suggestion to combine the conditions for `HTTPGet` and `TCPSocket` readiness probes. It’s cleaner now!

**Comment by LOADBC** [comment]
Hi @spolti , I have made the changes as per your feedback. Could you please review it again when you get a chance? Thank you!

**Comment by spolti** [comment]
@LOADBC thanks.

cc @yuzisun @terrytangyuan

**Comment by LOADBC** [comment]
@spolti can you approve the workflow

File: pkg/webhook/admission/pod/agent_injector.go [hunk 2]
```diff
@@ -238,12 +238,26 @@ func (ag *AgentInjector) InjectAgent(pod *v1.Pod) error {
...
+				readinessProbeJson, err := json.Marshal(readinessProbe)
+				if err != nil {
+					return fmt.Errorf("failed to marshal readiness probe: %w", err)
+				}
+
+				// Append the marshaled readiness probe as an environment variable for the agent container
+				agentEnvs = append(agentEnvs, v1.EnvVar{Name: "SERVING_READINESS_PROBE", Value: string(readinessProbeJson)})
+			} else if readinessProbe.Exec != nil {
+				// Skip ExecAction probes; do not inherit them
+				fmt.Printf("INFO: Exec readiness probe skipped for pod %s/%s\n", pod.Namespace, pod.Name)
```
Comment (sivanantha321) [inline pkg/webhook/admission/pod/agent_injector.go]: Can't we use a logger ?

Comment (LOADBC) [reply pkg/webhook/admission/pod/agent_injector.go]: Good point! @sivanantha321  I've taken your suggestion and added a logger to the code for improved consistency and observability, Please let me know if its okay.

**Review by sivanantha321 — APPROVED** [review]

**Pull request merged**